vulnerabilities/models.py

```diff
@@ -39,8 +42,48 @@ class Vulnerability(models.Model):
     VulnerabilityReference.
     """
 
-    cve_id = models.CharField(max_length=50, help_text="CVE ID", unique=True, null=True)
-    summary = models.TextField(help_text="Summary of the vulnerability", blank=True)
+    vulnerability_id = models.CharField(
+        max_length=50,
+        help_text="Unique vulnerability_id for a vulnerability: this is either a published CVE id"
+        " (as in CVE-2020-7965) if it exists. Otherwise this is a VulnerableCode-assigned VULCOID"
+        " (as in VULCOID-2021-01-23-15-12). When a vulnerability CVE is assigned later we replace"
+        " this with the CVE and keep the 'old' VULCOID in the 'old_vulnerability_id' field to "
+        "support redirection to the CVE id.",
+        unique=True,
+    )
+    old_vulnerability_id = models.CharField(
+        max_length=50,
+        help_text="empty if no  CVE else VC id",
+        unique=True,
+        null=True,
+    )
+    summary = models.TextField(
+        help_text="Summary of the vulnerability",
+        blank=True,
+    )
+
+    def save(self, *args, **kwargs):
+        if self.vulnerability_id:
+            return super().save(*args, **kwargs)
+        # Generate unique VULCOID
+        ie = None
+        for attempt in range(1, 11):
+            try:
+                self.vulnerability_id = self.generate_vulcoid()
+                # Using the context manager due to https://stackoverflow.com/a/23326971
+                with transaction.atomic():
```

Comment:
I am somewhat concerned about starting a transaction inside the save()

Reply:
I share you concern too. 
For some reason without that transaction block I get a :
`django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.`

There could be a deeper issue maybe ?

Reply:
Going with a transaction management approach here, you are asking for bigger issues than the one you are trying to solve initially.
If you want to go with the timestamp approach, I would simply add the microseconds (".%f") to the string to ensure unicity.

Reply:
@tdruez  that would solve the issue, but we'd have longer ids, @pombredanne  ^ would that be too ugly ?

Reply:
@tdruez  what would be cleaner way to do as @pombredanne  said 

> we retry until we can get an id that is unique in the rare cases where we have collisions

Reply:
> @sbs2001 but we'd have longer ids

Keep only 2 digits of the micro seconds then, you'll have something unique enough and only 1 char extra compared to the incremental id approach.

Reply:
@tdruez  that could make the collisions even rarer. But there would be a still (smaller) chance of collisions ? I am fine with whatever id we use. The code to deal with exceptions, especially the requirement of `transaction` is weird.

Reply:
I don't think it's likely that you will ever generate 2 vulcoid on the same micro second.
Encountering collisions with an arbitrary range of 10 integers is more likely.

Reply:
Current `strftime("%Y-%m-%d%H%M%S")` -> `VULCOID-2021-02-16084802`

Suggested `strftime("%Y%m%d-%H%M-%S%f")[:-4]` -> `VULCOID-20210216-0848-3008`

Reply:
Still collisions happen.

Reply:
What about keeping all the micro seconds? ie removing the [:-4]

Reply:
Alternatively, in place of the micro seconds, we could use a slice of a random UUID.

Reply:
> What about keeping all the micro seconds?

I used that approach while prototyping this whole thing, https://github.com/nexB/vulnerablecode/blob/787ab6e4778b50ddeed6f64146f3a6f58a4f57b2/vulnerabilities/models.py#L48  and it works ! 
It looks ugly but it works :)

Reply:
OK, let's be ugly but working then! Go for it.